presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,160 @@
+        // For testing we want strong references, otherwise the tests will be non-deterministic due to GC effects.
+        return new ClusterMemoryLeakDetector(new LeakedQueries(false));
+    }
+
+    private ClusterMemoryLeakDetector(LeakedQueries leakedQueries)
+    {
+        this.leakedQueries = requireNonNull(leakedQueries, "leakedQueries is null");
+    }
+
+    /**
+     * @param queryExecutionsSupplier All queries that the coordinator knows about.
+     * @param queryMemoryReservations The memory reservations of queries in the GENERAL cluster memory pool.
+     */
+    void checkForMemoryLeaks(Supplier<List<QueryInfo>> queryInfoSupplier, Map<QueryId, Long> queryMemoryReservations)
+    {
+        requireNonNull(queryInfoSupplier);
+        requireNonNull(queryMemoryReservations);
+
+        Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
+                .stream()
+                .filter(Objects::nonNull)
+                .filter(entry -> entry.getValue() > 0)
+                .filter(entry -> isLeaked(queryInfoSupplier, entry.getKey()))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+        if (!leakedQueryReservations.isEmpty()) {
+            log.warn("Memory leak detected. The following queries are already finished, " +
+                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
+        }
+
+        // keep numberOfLeakedQueries and leakedQueries in sync
+        synchronized (this) {
+            numberOfLeakedQueries.set(leakedQueryReservations.size());
+            leakedQueries.update(leakedQueryReservations.keySet());
+        }
+    }
+
+    private static boolean isLeaked(Supplier<List<QueryInfo>> queryInfoSupplier, QueryId queryId)
+    {
+        boolean queryExists = false;
+        QueryState queryState = null;
+        DateTime queryEndTime = null;
+
+        for (QueryInfo queryInfo : queryInfoSupplier.get()) {
+            if (queryInfo.getQueryId().equals(queryId)) {
+                queryExists = true;
+                queryState = queryInfo.getState();
+                queryEndTime = queryInfo.getQueryStats().getEndTime();
+            }
+        }
+
+        if (!queryExists) {
+            return true;
+        }
+
+        if (queryState.equals(RUNNING) || queryEndTime == null) {
+            return false;
+        }
+
+        return secondsBetween(queryEndTime, now()).getSeconds() >= DEFAULT_LEAK_CLAIM_DELTA_SEC;
+    }
+
+    synchronized boolean wasQueryLeaked(QueryId queryId)
+    {
+        return leakedQueries.contains(queryId);
+    }
+
+    int getNumberOfLeakedQueries()
+    {
+        return numberOfLeakedQueries.get();
+    }
+
+    static class LeakedQueries
+    {
+        private final Set<QueryId> leakedQueries;
+
+        LeakedQueries(boolean useWeakReferences)
+        {
+            this.leakedQueries = useWeakReferences ? newSetFromMap(new WeakHashMap<>()) : new HashSet<>();
```

Comment:
I'm not sure this works.  The actual `QueryId` instance has to be retained by someone else for the weak reference to remain valid.  There isn't any gaurentees on the retention of this object, so it could be cleared instantly.   Instead maybe we should use a time+size limited guava cache?

Reply:
What would happen if we used hard (normal) references always? `this.leakedQueries` collection contains elements only temporarily, until next refresh run, so it won't prevent objects from being GC'd for long time. What am i missing?

Reply:
I think we can use strong references, initially the initial version wasn't refreshing periodically, but the new version does that and we won't hold onto this state for too long anyways.